Add drag and drop reordering of pinned items

daltojohnso opened the pull request:
Related to #19758.
Resolves #19997.

This PR adds drag-and-drop sorting to pinned collection items. Questions, cards, and models are still split into separate lists, but you can click, drag, and drop within a section to move a given item either forward or backward to reorder the section.

This code uses `react-dnd` as all Collections code does. I tried using `react-sortable-hoc` to do this since parameter widget sorting is so much more pleasant, but `react-dnd` and `react-sortable-hoc` don't really play well together, and we are trying to avoid a separate "edit" mode right now.

I added a few new props to the existing `PinDropTarget` component,`itemModel`, `isFrontTarget`, and `isBackTarget`, all of which are used in the component's `canDrop` function to determine whether or not the drop target component is usable by whatever item you're dragging. `itemModel` is simple; you can't move a dashboard into the question section, etc. 

Concerning `isFrontTarget` and `isBackTarget`, you'll see in the `PinnedItemOverview` that I've added **two** `PinnedItemSortDropTarget` components next to each item component, one in front and one behind. This is so that when you are dragging an item we can add a little bit of styling to show the user where the item will be places upon dropping it.

**Testing**
1. Create a collection and pin questions, dashboards, models.
2. Click and drag pinned items and drop them in drop targets that appear on the page, either in front of or behind the dragged item.
3. Dropping an item should reorder the list, pushing other items back if you move the item forward or pushing other items forward if you move the item backward.

https://user-images.githubusercontent.com/13057258/151447249-3a6ae92a-8b43-4822-9e03-db38c323e267.mov

daltojohnso committed:
onDrop unneeded

daltojohnso committed:
basic impl of reordering pinned items

daltojohnso committed:
split cards and dashboards again

daltojohnso committed:
Improve pin drop target styling

daltojohnso committed:
Fix card drop target weirdness

mazameli commented on frontend/src/metabase/collections/components/PinnedItemOverview/PinnedItemOverview.tsx:
```diff
@@ -120,3 +187,42 @@ function PinnedItemOverview({
+      left: 0,
+      right: 0,
+      borderLeft: isFrontTarget
+        ? `4px solid ${color("brand-light")}`
```
```suggestion
        ? `4px solid ${color("brand")}`
```

I think making this `brand` here and on line 209 will make this stand out more, though I appeciate how you tried making this look more like my mockup initially.

daltojohnso committed:
brand-light border --> brand

daltojohnso committed:
pull out separate styled drop target component

daltojohnso committed:
replace style prop usage

daltojohnso committed:
reduce the z-index

kulyk commented:
DND works nicely and the code looks great 🎉 There are a few behaviors I believe we should discuss and address though

<details>
<summary>Accidentally triggering a drill though can result in a weird DND state</summary>
For some visualizations like line charts, you can drill through to some portion of data using mouse/trackpad/whatever like that:

![CleanShot 2022-01-28 at 14 53 00](https://user-images.githubusercontent.com/17258145/151550426-42421a8b-6021-4d77-af12-1b04f0a1f15a.gif)

Looks like this feature doesn't play nice with collection items DND. I've accidentally triggered that and ended up in a weird state:

https://user-images.githubusercontent.com/17258145/151550572-32b77084-3d11-40fa-88c8-a052238451aa.mov

Should we maybe disable this feature if we're on the collection page?
</details>

<details>
<summary>Query reruns and UI jumps when moving items</summary>
If there are pinned questions, they will be fully rerendered and their queries will run again once you move anything using DND (swap questions/models/dashboards order or move a thing to a different collection). The UI jump is pretty annoying and it can be a bad experience if you have some heavy queries pinned in a collection. Is it possible to avoid the rerenders and reruns at least when we're not moving questions? It feels wrong I need to rerun queries if I just move something to another collection.

https://user-images.githubusercontent.com/17258145/151550641-edeff16f-f8cb-493e-a528-ad759d29b0ea.mp4
</details>

<details>
<summary>Nip: pin/unpin help message</summary>
When a collection only has pinned items and you start dragging, there is a help banner saying "Drop items here to unpin them". Should we maybe display a similar message when all the items are unpinned? Otherwise, there is no way to pin items using DND until there is at least one pinned item in a group. That's just an idea though

![ux-improvement](https://user-images.githubusercontent.com/17258145/151550121-1dfa8157-f931-4d7f-b99e-f77a10524251.gif)
</details>

daltojohnso committed:
Remove prop that triggers drill functionality

daltojohnso committed:
Fix clobbering of pinned list on update

daltojohnso committed:
Fix card queries on every rerender

daltojohnso committed:
Add pin drop zone & split drop target components

daltojohnso commented:
Fixed all those things @kulyk -- thanks for the review.

Drilldown stuff was triggered by this stub function I was using to cheat and get highlighted title text in the visualization. Now, I'm using styled components to do it a little better: https://github.com/metabase/metabase/pull/19966/commits/de2b65a276e162b1796337600e14dd2d32b8c3f9

The items list clobbering was caused by the addition of this line (https://github.com/metabase/metabase/pull/19966/commits/5f85c3e2a4e4aca5401c17b6b89b3d90f4d36a7e) previously by me, but not sure why I added it now. Looks much better now.

I'm preventing card query reruns by storying the `question` instance I instantiate in `CollectionCardVisualization` so that `QuestionResultLoader` doesn't think it is getting a new question on every rerender: https://github.com/metabase/metabase/pull/19966/commits/61fd3e35fd715965bed8f7f06bda0134a38131ba

Added a pin drop zone here: https://github.com/metabase/metabase/pull/19966/commits/f6e88cf0f0ae661fc1660be509222e5b09a5e213.

daltojohnso committed:
lint fix

daltojohnso committed:
fix & add a few tests

mazameli commented:
The way you got rid of that reloading is just awesome.

What is this gray area at the top of the collections list meant to signify when dragging, though? I initially thought it was supposed to allow me to drag a pinned item there to unpin it but it doesn't seem to do anything. I would remove this unless there's a purpose I'm not seeing, because it actually distracts from the reordering indicator.

![image](https://user-images.githubusercontent.com/2223916/151632845-8ae0384c-2577-4d54-9393-da38673bba0b.png)

The new drop zone for pinning something could use a bit of love, too:

![image](https://user-images.githubusercontent.com/2223916/151633557-69b931ac-38a3-4b6b-8215-41ae3aad9287.png)

daltojohnso commented:
@mazameli -- seems like dropping in the gray header unpins a pinned item

daltojohnso commented:
@mazameli -- would you prefer I remove the pin drop zone and just make it so that you can drop anywhere in the overview section? Same for pinned items into the unpinned table; I think that was the intended effect of the gray table headers. I can probably pretty easily update it to apply an overlay with text. The drop zone styling was lazily taken from what currently exists when dragging an unpinned item 🙂

mazameli commented:
> would you prefer I remove the pin drop zone and just make it so that you can drop anywhere in the overview section?

Yes, I think that's preferable. We could even reuse the blue bar as an indicator.

>  Same for pinned items into the unpinned table; I think that was the intended effect of the gray table headers. I can probably pretty easily update it to apply an overlay with text.

Yeah, I'd definitely scrap the gray background behind the column headers. I don't think we even want a text overlay cause I think that finding out you can unpin something by dropping it back into the list is cool, but an optimization/shortcut for the explicit unpin action; and I don't want a text overlay distracting from the re-ordering indicators, since the only way to reorder is via dragging, and I think it may be a more common action than unpinning.

daltojohnso committed:
update pin drop zone styling to match sorting

daltojohnso commented:
Slapped together a quick implementation of this -- lmk what you think/want changed:

https://user-images.githubusercontent.com/13057258/151667809-716394c3-3ab4-419e-b81e-fb889ad2117c.mov

kulyk reviewed (approved):
🎉 🎉 🎉

daltojohnso commented:
I'm going to go ahead and merge this into `master` so that we can also get it into the 42 branch. @mazameli if there are additional UI tweaks you want made to the DND we can address in a separate PR. I'm mostly comfortable with that code now, so I can tweak things pretty quickly 🙂
